### DISCUSSION TITLE
tutorial-advanced/Air001_start

### DISCUSSION THREAD

**Discussion by giscus** [Air-duino/document#7] [comment]
# tutorial-advanced/Air001_start

连上开发板 使用串口连接上板子USB另一头的串口引脚，具体可以参考：使用串口下载 (https://wiki.luatos.com/chips/air001/mcu.html#id4) 如果购买了合宙的DAPLink调试器 如果是购买的小白板和合宙的DAPLink调试器，可以直接通过排针排母相连，方向如图所示 5 连接后，保证开发板有供电，开发板选择A...

https://arduino.luatos.com/tutorial-advanced/Air001_start.html

**Comment by liguichuan** [comment]
Arduino在Air001开发板上，第一个UART串口使用Serial.print()，那么要使用第二个UART串口使用的是什么？

**Reply by chenxuuu** [reply]
Serial1

**Reply by HalfSweet** [reply]
应当是Serial2，Serial1已经被默认分配给了全局的Serial

**Reply by liguichuan** [reply]
都不对，用之前还需要加多这一句：HardwareSerial UART2(PA_1,PA_0);

**Reply by HalfSweet** [reply]
之前的定义有问题，可能会导致误解，已经在0.4.0版本中进行修复

如何使用更多的uart，可以参考https://arduino.luatos.com/library/uart.html#airmcu-专用api

**Comment by liguichuan** [comment]
Air001如何在Arduino中获得：产品唯一身份标识码（UID）即Chip PID is: 0x04 0x40

**Reply by HalfSweet** [reply]
此功能暂不支持，如果您认为有必要，可以提一个相应的issue

**Comment by liguichuan** [comment]
Air001关于模拟EEPROM来存储用户需要保存的数据，有没有什么支持的库？

**Reply by HalfSweet** [reply]
直接`#include <EEPROM.h>` 即可，API与Arduino的标准库兼容

**Reply by liguichuan** [reply]
以下代码编译不报错，但也实现不了断电保存数据：
#include <EEPROM.h>
//创建一个共用体，用来存放ROM变量,断电保存
union eeprom_data
{
  double CF_CNT;//电量脉冲计数存储量。
  byte count[8];
};

eeprom_data CF;


int CF_address = 0;

void CF_Clean()
{
  // 让EEPROM的8字节内容全部清零
  for (int i = 0; i < 8; i++)
    EEPROM.write(i, 0);
}

void CF_Write()
{

    for (int i = CF_address; i < 8; i++)
    {
        EEPROM.write(i, CF.count[i]);
    }
}

void CF_Read()
{


    for (int i = CF_address; i < 8; i++)
    {
        CF.count[i] = EEPROM.read(i);
    }

}

方法使用：
            double CF_CNT = 1;
            CF_Read();
            delay(1);
            CF.CF_CNT = CF.CF_CNT + CF_CNT;
            CF_Write();

            Serial.print(":>[CF_CNT]");Serial.println(CF.CF_CNT);

**Reply by chenxuuu** [reply]
如有bug请上报仓库新建issue，此处仅用于讨论文档内容

**Comment by lllwan** [comment]
MACOS下使用daplink 通过arduino无法正常烧录：
 项目使用 15052 字节（45%）的程序存储空间。最大值为 32768 字节。
 个全局变量使用 1248 个字节（30%）的动态内存，剩下 2848 个字节用于局部变量。最大值为 4096 字节。
"/Users/my/Library/Arduino15/packages/AirM2M/tools/AirISP/0.1.4/AirISP-next" --chip "auto" --port "/dev/cu.usbmodem14102" --baud 115200 --before default_reset --after hard_reset write_flash -e 0x08000000 "/private/var/folders/7x/25_3v9cj6tv4h2w6c49crj4w0000gn/T/arduino/sketches/539A41C1A47C80D8855E6C6098D69035/sketch_jan20a.ino.hex"
Language not supported
Language not supported
AirISP version: 0.1.4
Opening serial port /dev/cu.usbmodem14102 ……
Connecting...___...___...___...___...___...___...___...
Failed to automatically enter boot mode, operation timed out, ending operation...
You can try to manually enter boot mode:
Unplug USB, hold the BOOT button, plug in USB, and retry the download operation.
After downloading, release the BOOT button and press RST to reset.
上传失败: 上传错误：exit status 1
